Source/APIClient/Authentication.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  AuthTokenRefresher.swift
+//  APIClient
+//
+//  Created by Mark Daigneault on 3/3/17.
+//  Copyright © 2017 Mark Daigneault. All rights reserved.
+//
+
+import Foundation
+import Intrepid
+
+public protocol AuthClient {
+    func login(email: String, password: String, completion: ((Result<String>) -> Void)?)
+}
+
+public protocol CredentialProviding {
+    var email: String? { get set }
+    var password: String? { get set }
+    var token: String? { get set }
+
+    var formattedToken: String? { get }
+}
+
+public extension CredentialProviding {
+    func authorizeRequest(_ request: inout URLRequest) {
+        request.setValue(formattedToken, forHTTPHeaderField: "Authorization")
+    }
+}
+
+public enum AuthTokenRefreshError: Error {
+    case MissingCredentials
+    case AuthenticationError(Error)
+}
+
+public class AuthTokenRefresher {
+
+    let apiClient: APIClient
+    let authClient: AuthClient
+    var credentialProvider: CredentialProviding
+
+    init(apiClient: APIClient, authClient: AuthClient, credentialProvider: CredentialProviding) {
+        self.apiClient = apiClient
+        self.authClient = authClient
+        self.credentialProvider = credentialProvider
+    }
+
+    func handleUnauthorizedRequest(request: URLRequest, completion: ((Result<Data?>) -> Void)?) {
+        guard
+            let email = credentialProvider.email,
+            let password = credentialProvider.password
+        else {
+            completion?(.failure(AuthTokenRefreshError.MissingCredentials))
+            return
+        }
+
+        authClient.login(email: email, password: password) { [weak self] result in
+            switch result {
+            case .success(let token):
+                var mutableRequest = request
+
+                self?.credentialProvider.token = token
+                self?.credentialProvider.authorizeRequest(&mutableRequest)
```

Comment:
Ahh I see. Unless we make another call to `Request.someRequest.urlRequest`, the `Authorization` header won't be re-set. I was thinking that that would automatically happen when we hit ` self?.apiClient.sendRequest(mutableRequest, completion: completion)` below, but it won't.

I suppose that if we _wanted_  to force the update through `Request.urlRequest`, we would pass a `Request` to `handleUnauthorizedRequest`, set the new token, and then general a new `urlRequest`. That would, however, reset anything we set "by hand" on the unauthorized request. 

So that seems like a worse option.